src/wayland/compositor/tree.rs

```diff
@@ -7,11 +7,15 @@ use super::{
     BufferAssignment, SurfaceAttributes, SurfaceData,
 };
 use std::{
+    any::Any,
     fmt,
-    sync::{atomic::Ordering, Mutex},
+    sync::{atomic::Ordering, Arc, Mutex},
 };
 use wayland_server::{backend::ObjectId, protocol::wl_surface::WlSurface, DisplayHandle, Resource};
 
+type CommitHook = dyn Fn(&mut dyn Any, &DisplayHandle, &WlSurface) + Send + Sync;
```

Comment:
Is there a reason why commit hook needs Sync but destroy does not?

Reply:
Because the hook is wrapped in an Arc, so we can clone the whole vector and release the lock.

Because Arcs are references, this means that technically multiple references could exist across multiple threads (even though they don't in this case), which is why `Arc` only implements `Send` if `T: Send + Sync`.
I don't think this additional bound is particularly limiting or hurtful, because it just affects captured variables, but we could work around that with a newtype and an unsafe `Send`-impl. But I don't really want to do that in case somebody in the future touches the code and breaks thread-safety accidentally.